### ISSUE TITLE
Feature: Option to start virtual Networks only without virtual Machines

### ISSUE THREAD

**Issue by uvsmtid** [mitchellh/vagrant#5923]
Please consider adding `--networks-only` option:
```
vagrant up --network-only
```

There are many scenarios when virtual network may be needed before virtual machines are created.

For example, automation stage when only physical host has to be accessed using IP address on virtual network first.

Origin: http://stackoverflow.com/q/31334008/441652

**Comment by sethvargo** [comment]
Hi @uvsmtid 

Can you please explain your use case a bit more?

**Comment by uvsmtid** [comment]
## Stripped down example ##

* Let's say I use multi-staged deployment.
* And virtual machines need a services on virtual network (i.g. Salt, Jenkins, ...) before provisioning.
* This service is supposed to be started and configured first (possibly via known IP address through HTTP-based APIs).

The natural sequence is to create network first, start the service, configure it, create virtual machine with provisioning (which depends on the services).

## The point ##

Virtual network configuration is _predictable_ - IP addresses can be controlled (unlike any physical networks which depends on target environment). And Vagrant already isolates from details of creating required networks for various providers.

Splitting process of (A) = creating virtual networks and (C) = creating virtual machines allows (B) = configuring any infrastructure services using _predictable_ IP networking. Unified sequence (A), (B), (C).

**Comment by mitchellh** [comment]
While this sounds useful, I don't think this is useful in Vagrant itself. Vagrant is a tool focused on development environments. "Multi-staged deployment" of software is not a goal of Vagrant and we dont' want to build features into Vagrant to do that. 

There are arguments to be made that some applications need this feature, but it is so rare (this is the first time in 6 years we've heard it) that I'd say that this isn't needed enough to go into Vagrant. This should fit well in a plugin. 

For a real multi-stage deployment tool, check out Terraform. It was made for this.

**Comment by uvsmtid** [comment]
It makes sense if it's rare.

However, I'm using it for development only. It's development of deployment sequence. Or it's a continuous integration platform re-instantiating everything from zero in VMs managed by Vagrant (Vagrant is the simplest way to run it with any convenient provider).

## Workaround ##

A resource-intensive way is to create VM and destroy it just for the sake of creating networks:
```shell

# Instantiate network only. Networks stay configured after machines are destroyed.
vagrant up --no-provision && vagrant destroy

# ...
# Do steps required to configure network services.
# ...

#  Repeat machine creation with provisioning (when required network services are already up).
vagrant up --provision
```

**Comment by uvsmtid** [comment]
Surprisingly, it's exactly one year after my last comment.

And I just had more struggle with workarounds. Is it good time to review this idea again? Let me try again. :)

### Use case revisited ### 

The most convenient place to provide additional services and resources (e.g. per developer) is the host OS (e.g. developer's physical machine). And these services cannot be started on the virtual IP address until the virtual network has been created.

If anyone uses virtual network for testing and simulation (very common obvious case) and runs automated application deployment from scratch in their continuous application platform, then preparing the virtual network and all the services _before_ starting VMs must be a very convenient option to have.

### Cohesion ###

Vagrant is the only right way to propagate "network only option". Doing it outside of Vagrant is like creating another separate `network-vagrant` tool which abstracts various virtualization providers via uniform network creation interface. The networks are already fully defined in `Vagrantfile`. Does it make any sense? Vagrant already runs all the code - any custom tool would be a pathetic reimplementation of this existing functionality.

Why would `vagrant snapshot` exists if not because snapshotting is a common feature to many providers? It is natural to provide wrapper (CLI) front-end (which Vagrant is valuable for). Doesn't this approach extrapolate on network creation as well?

In fact, `vagrant` itself does not need to do much - it just have to recognize that it manages virtual network creation anyway and it is essentially different thing from VM creation. If there was a simple option (or subcommand) to let providers know, the support is up to the individual provider.

Why not allow the (CLI) front-end to propagate down such user request to the providers now? The actual support may just be experimental at the moment.
